sdk/storage/Azure.Storage.Blobs/tests/TransactionalHashingTests.cs

```diff
@@ -0,0 +1,653 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using System.Threading.Tasks;
+using Azure.Storage.Blobs.Models;
+using Azure.Storage.Blobs.Specialized;
+using Azure.Storage.Test.Shared;
+using NUnit.Framework;
+
+namespace Azure.Storage.Blobs.Tests
+{
+    public class TransactionalHashingTests : BlobTestBase
+    {
+        #region Test Arg Definitions
+        private const long DefaultDataSize = Constants.KB;
+        private static IEnumerable<HttpRange> DefaultDataHttpRanges
+        {
+            get
+            {
+                yield return new HttpRange(0, 512);
+                yield return new HttpRange(256, 512);
+                yield return new HttpRange(512, 512);
+            }
+        }
+
+        private static IEnumerable<(int DataSize, int BufferSize)> StorageStreamDefinitions
+        {
+            get
+            {
+                yield return (Constants.KB, Constants.KB);
+            }
+        }
+        #endregion
+
+        public TransactionalHashingTests(bool async, BlobClientOptions.ServiceVersion serviceVersion)
+            : base(async, serviceVersion, null /* RecordedTestMode.Record /* to re-record */)
+        {
+        }
+
+        #region DownloadContent
+        [Test, Combinatorial]
+        public async Task DownloadContentSuccessfulHashVerification(
+            [Values(TransactionalHashAlgorithm.MD5, TransactionalHashAlgorithm.StorageCrc64)] TransactionalHashAlgorithm algorithm,
+            [ValueSource("DefaultDataHttpRanges")] HttpRange range)
+        {
+            await using DisposingContainer test = await GetTestContainerAsync();
+
+            var blobName = GetNewBlobName();
+            await TransactionalHashingTestSkeletons.TestDownloadSuccessfulHashVerificationAsync(
```

Comment:
Would it possible to simplify this test class?  I think the lambdas can be non-intuitive, and it could make these tests difficult to maintain.

Reply:
The issue is that we need to test the exact same functionality across BlobClient, BlockBlobClient, AppendBlobClient, PageBlobClient, DataLakeFileClient, and ShareFileClient. The tests aren't the most intuitive things in the first place and cloning a dozen-ish of those tests 6 times is a maintainability nightmare. Genericizing the tests with lambdas to get the correct clients seemed to be the best approach. Perhaps there are optimizations to what the lambdas are but we need to inject resource creation and the operation to test.

Reply:
At minimum, I think we need to add a lot of additional inline comments in the tests and shared test code to explain what is going on here.  These are some of the most complicated unit tests we have, and I'm concerned with maintainability if it isn't generally known how these work.

Reply:
I will extensively docstring/inline the shared code as well as put some "why" descriptions on the methods/lambdas passed in.

Reply:
This is not a bad idea to create a template for hashing tests, but I'm not sure if current implementation reduces the codebase (i.e. I think that if we just multiplied the logic (copy-paste) the volume would be similar or even a bit smaller).

How about something like:
```
public abstract class TransactionalHashingTest {
     public TransactionalHashingTest (
      Delegate1 func1,
      Delegate2 func2,
      ...
     ) {
        this.func1 = func1;
        this.func2 = func2;
     }


     [Test]
     public async Task TestSomething() {
        await func1();
       
     }
     
     [Test]
     public async Task TestSomethingElse(){
       await func2();
     }
}

// and then have

public class BlockBlobHashingTests : TransactionalHashingTest {

   public BlockBlobHashingTests : base (func1, func2) {
   
   }

}

public class AppendBlobHashingTests : TransactionalHashingTest {

   public AppendBlobHashingTests : base (func1, func2) {
   
   }

}


public class ShareFileTransactionalHashingTest : TransactionalHashingTest {

   public ShareFileTransactionalHashingTest : base (func1, func2) {
   
   }

}

```

Reply:
@jaschrep-msft In this particular case , since we have an inheritance tree, these delegates might be abstract virtual methods perhaps.